niworkflows/interfaces/cifti.py

```diff
@@ -272,7 +134,266 @@ class CiftiNameSource(SimpleInterface):
+        bold_img = resample_to_img(bold_img, label_img)
+
+    bold_data = bold_img.get_fdata(dtype='float32')
+    timepoints = bold_img.shape[3]
+    label_data = np.asanyarray(label_img.dataobj).astype('int16')
+
+    # Create brain models
+    idx_offset = 0
+    brainmodels = []
+    bm_ts = np.empty((timepoints, 0))
+
+    for structure, labels in CIFTI_STRUCT_WITH_LABELS.items():
+        if labels is None:  # surface model
+            model_type = "CIFTI_MODEL_TYPE_SURFACE"
+            # use the corresponding annotation
+            hemi = structure.split('_')[-1]
+            # currently only supports L/R cortex
+            surf = nb.load(bold_surfs[hemi == "RIGHT"])
+            surf_verts = len(surf.darrays[0].data)
+            if annotation_files[0].endswith('.annot'):
+                annot = nb.freesurfer.read_annot(annotation_files[hemi == "RIGHT"])
+                # remove medial wall
+                medial = np.nonzero(annot[0] != annot[2].index(b'unknown'))[0]
+            else:
+                annot = nb.load(annotation_files[hemi == "RIGHT"])
+                medial = np.nonzero(annot.darrays[0].data)[0]
+            # extract values across volumes
+            ts = np.array([tsarr.data[medial] for tsarr in surf.darrays])
+
+            vert_idx = ci.Cifti2VertexIndices(medial)
+            bm = ci.Cifti2BrainModel(
+                index_offset=idx_offset,
+                index_count=len(vert_idx),
+                model_type=model_type,
+                brain_structure=structure,
+                vertex_indices=vert_idx,
+                n_surface_vertices=surf_verts
+            )
+            idx_offset += len(vert_idx)
+            bm_ts = np.column_stack((bm_ts, ts))
+        else:
+            model_type = "CIFTI_MODEL_TYPE_VOXELS"
+            vox = []
+            ts = None
+            for label in labels:
+                ijk = np.nonzero(label_data == label)
+                if ijk[0].size == 0:  # skip label if nothing matches
+                    continue
+                ts = (bold_data[ijk] if ts is None
+                      else np.concatenate((ts, bold_data[ijk])))
+                vox += [[ijk[0][ix], ijk[1][ix], ijk[2][ix]]
+                        for ix, row in enumerate(ts)]
+
+            vox = ci.Cifti2VoxelIndicesIJK(vox)
+            bm = ci.Cifti2BrainModel(
+                index_offset=idx_offset,
+                index_count=len(vox),
+                model_type=model_type,
+                brain_structure=structure,
+                voxel_indices_ijk=vox
+            )
+            idx_offset += len(vox)
+            bm_ts = np.column_stack((bm_ts, ts.T))
+        # add each brain structure to list
+        brainmodels.append(bm)
+
+    # add volume information
+    brainmodels.append(
+        ci.Cifti2Volume(
+            bold_img.shape[:3],
+            ci.Cifti2TransformationMatrixVoxelIndicesIJKtoXYZ(-3, bold_img.affine)
+        )
+    )
```

Comment:
I believe here's where you sample the subcortical structures from the BOLD data. Now that we have learned so much from nitransforms, you can do this step without the resampling above.

The pseudocode would be:

```
ijk_labels = np.where(label_data > 0)  # I don't remember if you need to set an upper bound threshold
ijk_bold = np.linalg.inv(bold_img.affine).dot(label_img.affine.dot(ijk_labels))  # You'll need to take care of ijk_labels, add a ones column and probably transpose here.
bold_values = scipy.ndimage.map_coordinates(bold_data, ijk_bold, order=3)
```

Reply:
this looks really cool, though i'm inclined to leave this as is for now - we will likely address this and other external resamplings once nitransforms is incorporated into the code base